sdk/lib/rtl/rangelist.c

```diff
@@ -47,237 +509,675 @@ typedef struct _RTL_RANGE_ENTRY
+        DPRINT("RtlpDeleteFromMergedRange: Add %p to TmpList\n", CurrentRtlEntry);
+
+        Status = RtlpAddRange(&TmpList, CurrentRtlEntry, RTL_RANGE_LIST_ADD_IF_CONFLICT);
+
+        if (!NT_SUCCESS(Status))
+        {
+            if (Status != STATUS_INSUFFICIENT_RESOURCES)
+            {
+                DPRINT1("RtlpDeleteFromMergedRange: Status %X\n", Status);
+                ASSERT(Status == STATUS_INSUFFICIENT_RESOURCES);
+            }
+            else
+            {
+                DPRINT1("RtlpDeleteFromMergedRange: STATUS_INSUFFICIENT_RESOURCES\n");
+            }
+
+            CurrentRtlEntry = RtlpEntryFromLink(TmpList.Flink);
+            NextRtlEntry = RtlpEntryFromLink(CurrentRtlEntry->ListEntry.Flink);
+
+            while (&TmpList != &CurrentRtlEntry->ListEntry)
+            {
+                Status = RtlpAddToMergedRange(RtlEntry, CurrentRtlEntry, RTL_RANGE_LIST_ADD_IF_CONFLICT);
+                ASSERT(NT_SUCCESS(Status));
+
+                CurrentRtlEntry = NextRtlEntry;
+                NextRtlEntry = RtlpEntryFromLink(CurrentRtlEntry->ListEntry.Flink);
+            }
+
+            return RtlpAddToMergedRange(RtlEntry, MergedRtlEntry, RTL_RANGE_LIST_ADD_IF_CONFLICT);
+        }
+
+        CurrentRtlEntry = NextRtlEntry;
+        NextRtlEntry = RtlpEntryFromLink(NextRtlEntry->ListEntry.Flink);
+    }
+
+    if (IsListEmpty(&TmpList))
+    {
+        DPRINT("RtlpDeleteFromMergedRange: IsListEmpty(&TmpList)\n");
+        RemoveEntryList(&RtlEntry->ListEntry);
     }
     else
     {
-         //Previous = NULL;
-        Entry = RangeList->ListHead.Flink;
-        while (Entry != &RangeList->ListHead)
+        PLIST_ENTRY Flink;
+        PLIST_ENTRY Blink;
+
+        DPRINT("RtlpDeleteFromMergedRange: Add entry from TmpList\n");
+
+        Blink = RtlEntry->ListEntry.Blink;
+        Flink = RtlEntry->ListEntry.Flink;
+
+        Blink->Flink = TmpList.Flink;
+        TmpList.Flink->Blink = Blink;
+
+        Flink->Blink = TmpList.Blink;
+        TmpList.Blink->Flink = Flink;
+    }
+
+    //ExFreeToPagedLookasideList(&RtlpRangeListEntryLookasideList, MergedRtlEntry);
+    RtlpFreeMemory(MergedRtlEntry, 'elRR');
+    //ExFreeToPagedLookasideList(&RtlpRangeListEntryLookasideList, RtlEntry);
+    RtlpFreeMemory(RtlEntry, 'elRR');
+
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS
+NTAPI
+RtlDeleteOwnersRanges(
+    _In_ PRTL_RANGE_LIST RangeList,
+    _In_ PVOID Owner)
+{
+    PRTLP_RANGE_LIST_ENTRY RtlEntry;
+    PRTLP_RANGE_LIST_ENTRY NextRtlEntry;
+    PRTLP_RANGE_LIST_ENTRY MergedRtlEntry;
+    PRTLP_RANGE_LIST_ENTRY MergedNextEntry;
+    NTSTATUS Status = STATUS_SUCCESS;
+
+    PAGED_CODE_RTL();
+    ASSERT(RangeList);
+
+    DPRINT("RtlDeleteOwnersRanges: RangeList %p, Owner %p, [%X]\n", RangeList, Owner, RangeList->Count);
+
+START:
+
+    RtlEntry = RtlpEntryFromLink(RangeList->ListHead.Flink);
+    NextRtlEntry = RtlpEntryFromLink(RtlEntry->ListEntry.Flink);
+
+    while (&RtlEntry->ListEntry != &RangeList->ListHead)
```

Comment:
Here it also makes sense to use `for` loops. Should be possible to replace `goto START;` with `continue` after that

Reply:
Please show your version so that I understand what you mean.

Reply:
Since you're obtaining a list entry from the link you can wrap that chunk of code into a loop into something like this as an example:
```c
for (RtlEntry = RtlpEntryFromLink(RangeList->ListHead.Flink);
     RtlEntry->ListEntry != &RangeList->ListHead;
     RtlEntry->ListEntry = RtlEntry->ListEntry->Flink)
{
   if (MergedRtlEntry->Allocated.Owner == Owner)
   {
    [...]
    continue;
   }
}
```

Reply:
I meant RtlDeleteOwnersRanges(). Here's what I get.
Please offer your variants for getting rid of "START".
Or am I misunderstanding? 
``` c
RtlDeleteOwnersRanges(..)
{

START:

    for (RtlEntry = RtlpEntryFromLink(RangeList->ListHead.Flink),
         NextRtlEntry = RtlpEntryFromLink(RtlEntry->ListEntry.Flink);
         &RtlEntry->ListEntry != &RangeList->ListHead;
         RtlEntry = NextRtlEntry,
         NextRtlEntry = RtlpEntryFromLink(RtlEntry->ListEntry.Flink))
    {
        if (RtlEntry->PrivateFlags & RTLP_ENTRY_IS_MERGED)
        {
            for (MergedRtlEntry = RtlpEntryFromLink(RtlEntry->Merged.ListHead.Flink),
                 MergedNextEntry = RtlpEntryFromLink(MergedRtlEntry->ListEntry.Flink);
                 &MergedRtlEntry->ListEntry != &RtlEntry->Merged.ListHead;
                 MergedRtlEntry = MergedNextEntry,
                 MergedNextEntry = RtlpEntryFromLink(MergedRtlEntry->ListEntry.Flink))
            {
                if (MergedRtlEntry->Allocated.Owner == Owner)
                {
                    [..]
                    goto START;
                }
            }
        }
        else if (RtlEntry->Allocated.Owner == Owner)
        {
            [..]
        }
    }

    return Status;
}
```